sw/device/lib/dif/dif_alert_handler.h

```diff
@@ -0,0 +1,649 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_DIF_DIF_ALERT_HANDLER_H_
+#define OPENTITAN_SW_DEVICE_LIB_DIF_DIF_ALERT_HANDLER_H_
+
+/**
+ * @file
+ * @brief <a href="/hw/ip/alert_handler/doc/">Alert Handler</a> Device Interface
+ * Functions
+ */
+
+#include <stddef.h>
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_warn_unused_result.h"
+
+// Header Extern Guard (so header can be used from C and C++)
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * Generic enable/disable enumeration.
+ */
+typedef enum dif_alert_handler_enable {
+  kDifAlertHandlerEnabled,
+  kDifAlertHandlerDisabled,
+} dif_alert_handler_enable_t;
+
+/**
+ * An alert class.
+ *
+ * An alert class roughly specifies how to deal with an alert. The class
+ * determines which interrupt handler is fired for an alert, as well as the
+ * fine-grained details of the escalation policy, for when the processor
+ * fails to respond to an alert quickly enough.
+ *
+ * Alert classes serve as the alert handler's IRQ types. There is one IRQ for
+ * each class.
+ */
+typedef enum dif_alert_handler_class {
+  kDifAlertHandlerClassA,
+  kDifAlertHandlerClassB,
+  kDifAlertHandlerClassC,
+  kDifAlertHandlerClassD,
+} dif_alert_handler_class_t;
+
+/**
+ * An alert, identified by a numeric id.
+ *
+ * Alerts are hardware-level events indicating that something catastrophic
+ * has happened. The alert handler consumes alerts, classifies them to a
+ * particular `dif_alert_handler_class_t`, and uses policy information attached
+ * to that class to handle it.
+ *
+ * The number of alerts is configurable at hardware-synthesis time, and is
+ * specified by the software when initializing a `dif_alert_handler_t`.
+ */
+typedef uint32_t dif_alert_handler_alert_t;
```

Comment:
> One obvious approach that trades off reusability for readability would be defining an enum that contains both local and non-local alerts.

The problem boils down to "the alert handler doesn't actually know non-local alerts are". This is much how like the PLIC doesn't know what an interrupt is: for this reason, the PLIC doesn't have a giant enum of interrupts like it used to. However, the difference is that the alert handler does know about its *own* alerts, which in the HW design are handled separately.

What I've done here is done to mirror the HW design, which I think is the simplest design, especially for someone reading the comportability spec.

@msfschaffner explained a lot of these details to me. Do you have input?

Reply:
Sure, but how about the second alternative? I guess I don't see why clients have to use different functions for local and non-local alerts. Is there a particular reason?

Reply:
I'm not sure I like the second alternative. It adds another layer of abstraction over the hardware, where right now, the value of an alert_t is exactly that specified at the top-level. Now, callers need to know to add +4 to those values. I'm a bit uncomfortable about hiding this hardware detail from callers in a way that isn't actually invisible.

I think that it makes more sense to ask HW if they're willing to merge the two sets of registers, so that the first four alerts are defined to be the alert handler's own local alerts.

Reply:
Power manager has a similar structure, i.e. wake-up and reset sources. You may want to take a look at #3142, basically a struct w/ a tag.

Reply:
Indeed, we've separated out the internal alerts on purpose since this aligns nicely with how the HW design is constructed internally, and it makes it clear that these are of a different origin / type than the remaining "external" alerts that originate somewhere else in the chip. 

While it would be possible to merge these alerts into the same set of registers, I fear that this may have quite some impact on DV at this point.